Update square_dist_scalar_mult

jdemel opened the pull request:
This kernel has several problems. At least on my machines, the AVX version was slower than the generic one. Also, I did some refactoring to give this more structure.
Although, I haven't double checked, this should also fix #321 .
I wanted to add AVX2 meta intrinsics for this kernel but since #321 came up, I decided to open a PR now.

jdemel committed:
kernel: Refactor square_dist_scalar_mult

The 'square_dist_scalar_mult' kernel could be very useful for symbol
demapping. Currently, the AVX implementation seems to be slower than the
generic one. Also, the code is hard to read and needs refactoring.
Especially, move common intrinsics to `include/*_intrinsics.h` in order
to have a simple way to implement aligned and unaligned kernels.

jdemel committed:
square_dist_scalar_mult: Speed-Up AVX, Add unaligned

This patch includes 2 things. First, some refactoring to have common
functions as intrinsics in `include/volk/*_intrinsics.h`. Second, I
added unaligned kernels for AVX and SSE3.

Finally, at least on my machine, the existing AVX kernel was
slower than the generic kernel. I replaced it with my implementation.
Results shown below for, what I think, are realistic function calls as
part of a demapper.

`volk_profile -n -R square_dist_scalar_mult -v 1024 -i 10000000`
a_avx completed in 2355.5 ms (was: > 4500ms)
a_sse3 completed in 3561.28 ms
generic completed in 4549.62 ms
u_avx completed in 3271.98 ms
u_sse3 completed in 3657.44 ms

jdemel committed:
square_dist_scalar_mult: refactor AVX2 kernel

Rename variables and update structure for better readibility.

rear1019 commented on kernels/volk/volk_32fc_x2_s32f_square_dist_scalar_mult_32f.h:
```diff
@@ -79,69 +79,80 @@
 static inline void
-volk_32fc_x2_s32f_square_dist_scalar_mult_32f_a_avx2(float* target, lv_32fc_t* src0, lv_32fc_t* points,
-                                                     float scalar, unsigned int num_points)
+volk_32fc_x2_s32f_square_dist_scalar_mult_32f_a_avx2(float* target, lv_32fc_t* src0, 
```
Should be applied to the unaligned version as well.

jdemel replied:
I assume you refer to formatting?
Originally, I intended to add meta intrinsics for AVX2 too before i issue a PR but then an issue with this kernel came up and I decided to issue this PR now. I'll add the meta intrinsic and do the formatting.

rear1019 replied:
Yes, that’s what I am talking about. I missed that you already intended to do so.

rear1019 commented on kernels/volk/volk_32fc_x2_s32f_square_dist_scalar_mult_32f.h:
```diff
@@ -368,26 +316,21 @@ volk_32fc_x2_s32f_square_dist_scalar_mult_32f_generic(float* target, lv_32fc_t*
 static inline void
-volk_32fc_x2_s32f_square_dist_scalar_mult_32f_u_avx2(float* target, lv_32fc_t* src0, lv_32fc_t* points,
-                                                     float scalar, unsigned int num_points)
+volk_32fc_x2_s32f_square_dist_scalar_mult_32f_u_avx2(float* target, lv_32fc_t* src0, 
```
Changes from aligned version missing.

jdemel replied:
Yeah. I'll add them as soon as the aligned version is reworked.

rear1019 commented on kernels/volk/volk_32fc_x2_s32f_square_dist_scalar_mult_32f.h:
```diff
@@ -150,214 +161,151 @@ volk_32fc_x2_s32f_square_dist_scalar_mult_32f_a_avx2(float* target, lv_32fc_t* s
-    *target = scalar * (lv_creal(diff) * lv_creal(diff) + lv_cimag(diff) * lv_cimag(diff));
-    ++target;
-    ++points;
+volk_32fc_x2_s32f_square_dist_scalar_mult_32f_a_avx(float *target, lv_32fc_t *src0, 
```
The AVX version is now faster, tested with an Ivy Bridge (mobile) and Ryzen (Zen). On Ivy Bridge it is now faster than generic (with the old code the AVX version is as fast as generic or only slightly faster). On Ryzen the old version was already faster than genetic and now got even faster (but still slightly slower than AVX2).

jdemel replied:
On my SandyBridge laptop, the generic version was actually faster. That's how I came across this kernel.

rear1019 replied:
The AVX version is now *not* slower than the SSE3 version as well – at least on my Ryzen (Zen) machine it is slightly faster (no suprise given that Zen has only 128 Bit execution units). I do not have the numbers with Ivy Bridge at hand).

rear1019 commented on include/volk/volk_avx_intrinsics.h:
```diff
@@ -68,6 +68,19 @@ _mm256_magnitude_ps(__m256 cplxValue1, __m256 cplxValue2){
 }
 
+static inline __m256
+_mm256_scaled_norm_dist_ps(const __m256 symbols0, const __m256 symbols1, const __m256 points0, const __m256 points1, const __m256 scalar){
```
LGTM

rear1019 commented on include/volk/volk_sse3_intrinsics.h:
```diff
@@ -61,4 +61,17 @@ _mm_magnitude_ps_sse3(__m128 cplxValue1, __m128 cplxValue2){
 }
 
+static inline __m128
+_mm_scaled_norm_dist_ps_sse3(const __m128 symbols0, const __m128 symbols1, const __m128 points0, const __m128 points1, const __m128 scalar){
```
LGTM

rear1019 commented on kernels/volk/volk_32fc_x2_s32f_square_dist_scalar_mult_32f.h:
```diff
@@ -79,69 +79,80 @@
+    xmm_points1 = _mm256_load_ps((float*)(points + 4));
 
     xmm4 = _mm256_hadd_ps(xmm6, xmm7);
     xmm4 = _mm256_permutevar8x32_ps(xmm4, idx);
```
Just a rant: At first sight I stumbled a little and wondered why this is required. But than I remembered what a pain and how broken AVX is if you deal with complex numbers. I really envy NEON for having ``vld2``.

jdemel replied:
`_mm256_permutevar8x32_ps` is actually the reason the AVX2 kernel is faster than the AVX one. Before that intrinsic was added, there was a good chance you end up with an AVX kernel that is slower than its SSE counterpart.

rear1019 replied:
I am aware of that. But I better not digress any further…

rear1019 commented on kernels/volk/volk_32fc_x2_s32f_square_dist_scalar_mult_32f.h:
```diff
@@ -79,69 +79,80 @@
+
+
+static inline void
+calculate_scaled_distances(float* target, const lv_32fc_t symbol, const lv_32fc_t* points,
```
LGTM

rear1019 reviewed (changes requested):
Changed to the aligned AVX2 version should be applied to the unaligned AVX2 version as well.

rear1019 commented on kernels/volk/volk_32fc_x2_s32f_square_dist_scalar_mult_32f.h:
```diff
@@ -79,69 +79,80 @@
+  const __m256i idx = _mm256_set_epi32(7,6,3,2,5,4,1,0);
+  
+  xmm_points0 = _mm256_load_ps((float*)points);
+  xmm_points1 = _mm256_load_ps((float*)(points + 4));
```
These loads must not performed if ``num_points`` is smaller than 8 – out-of-bound memory access. Some day, on some machine this will fail with a SEGFAULT. Move into the loop below.

jdemel replied:
That's a good point. I'll do this.

rear1019 commented on kernels/volk/volk_32fc_x2_s32f_square_dist_scalar_mult_32f.h:
```diff
@@ -150,214 +161,151 @@ volk_32fc_x2_s32f_square_dist_scalar_mult_32f_a_avx2(float* target, lv_32fc_t* s
+  const __m128 xmm_scalar = _mm_load1_ps(&scalar);
+  
+  xmm_points0 = _mm_load_ps((float*)points);
+  xmm_points1 = _mm_load_ps((float*)(points + 2));
```
These loads must not performed if ``num_points`` is smaller than 4 – out-of-bound memory access. Some day, on some machine this will fail with a SEGFAULT. Move into the loop below.

rear1019 commented on kernels/volk/volk_32fc_x2_s32f_square_dist_scalar_mult_32f.h:
```diff
@@ -150,214 +161,151 @@ volk_32fc_x2_s32f_square_dist_scalar_mult_32f_a_avx2(float* target, lv_32fc_t* s
+                                            xmm_points1, xmm_scalar);
 
+  _mm_store_ps(target, xmm_result);
   target += 4;
```
Will not be required if the loads outside the loop are merged into the loop.

rear1019 commented on kernels/volk/volk_32fc_x2_s32f_square_dist_scalar_mult_32f.h:
```diff
@@ -444,18 +387,117 @@ volk_32fc_x2_s32f_square_dist_scalar_mult_32f_u_avx2(float* target, lv_32fc_t* s
+  const __m128 xmm_scalar = _mm_load1_ps(&scalar);
+  
+  xmm_points0 = _mm_loadu_ps((float*)points);
+  xmm_points1 = _mm_loadu_ps((float*)(points + 2));
```
Load outside of loop, see above.

rear1019 reviewed (changes requested):
I have updated my review. Missed some out-of-bound memory accesses (which were already present in the old code).

jdemel reviewed (commented):
Thanks for your comments @rear1019

michaelld commented:
What's the status of this PR? Looks like there's still some work to be done ...

jdemel commented:
Yeah. Unfortunately. I hope I'll manage to make progress soon.

jdemel committed:
kernel: create AVX2 meta intrinsics

With this commit the `square_dist_scalar_mult` kernel uses meta
intrinsics for AVX2 as well.

Potential out-of-bounds memory access problems are fixed as well.

jdemel commented:
I just updated this PR. All the things I wanted to put in are finished now. Also, I fixed potential out-of-bounds-memory-access issues. Thanks for pointing this out @rear1019 .

Also, I don't see any warnings with regards to comparing signed with unsigned anymore. I hope I fixed all of them.

I rebased this PR to the latest master. Just to be sure.

michaelld reviewed (approved):
LGTM & passes test locally. I do find it interesting the results:
```
test 98
    Start  98: qa_volk_32fc_x2_s32f_square_dist_scalar_mult_32f

98: Test command: /bin/sh "/opt/local/var/macports/build/_opt_sources_MacPorts_ports_github_macports_science_volk/volk-devel/work/build/lib/volk_32fc_x2_s32f_square_dist_scalar_mult_32f_test.sh" "/opt/local/var/macports/build/_opt_sources_MacPorts_ports_github_macports_science_volk/volk-devel/work/build/lib"
98: Test timeout computed to be: 10000000
98: RUN_VOLK_TESTS: volk_32fc_x2_s32f_square_dist_scalar_mult_32f(131071,1)
98: a_avx2 completed in 0.152 ms
98: a_avx completed in 0.146 ms
98: a_sse3 completed in 0.154 ms
98: generic completed in 0.143 ms
98: u_avx2 completed in 0.113 ms
98: u_avx completed in 0.061 ms
98: u_sse3 completed in 0.063 ms
98: Best aligned arch: u_avx
98: Best unaligned arch: u_avx
2/3 Test  #98: qa_volk_32fc_x2_s32f_square_dist_scalar_mult_32f ...   Passed    0.04 sec
```
Wondering why u_* is in general so much faster than a_* equivalent ...

michaelld commented:
@rear1019 any thoughts from you?

michaelld commented:
CI looks good. Not sure why the BB versions fail, but they have nothing to do with this PR.

jdemel commented:
> 98: Test timeout computed to be: 10000000
> 98: RUN_VOLK_TESTS: volk_32fc_x2_s32f_square_dist_scalar_mult_32f(131071,1)
> 98: a_avx2 completed in 0.152 ms
> 98: a_avx completed in 0.146 ms
> 98: a_sse3 completed in 0.154 ms
> 98: generic completed in 0.143 ms
> 98: u_avx2 completed in 0.113 ms
> 98: u_avx completed in 0.061 ms
> 98: u_sse3 completed in 0.063 ms
> 98: Best aligned arch: u_avx
> 98: Best unaligned arch: u_avx

For reliable results I run `volk_profile -n -R square_dist_scalar_mult -v 1024 -i 10000000`. Or even more 
iterations. What's your hardware setup? On my machines a_avx2 is slightly faster than a_avx. The unaligned versions are way slower.

rear1019 commented on kernels/volk/volk_32fc_x2_s32f_square_dist_scalar_mult_32f.h:
```diff
@@ -79,69 +79,73 @@
-    xmm4 = _mm256_mul_ps(xmm4, xmm8);
-
-    _mm256_store_ps(target, xmm4);
+    __VOLK_PREFETCH(points);
```
I wonder whether this results in any speed gain (this is not the case on my Ivy Bridge machine). Given that the memory access pattern is extremely regular I think that the CPU will prefetch anyway. Skimming through AMD’s and Intel’s optimization guides indicates that meaningful usage of ``prefetch`` is not trivial. See also Agner Fog’s optimization guide [1, chapter 12.14]:

> Automatic data prefetching often works better than explicit data prefetching when
the data access pattern is sufficiently regular.
> […]
> A good deal of experimentation is often needed to find the optimal prefetching
strategy for a program that accesses data in an irregular manner.

[1] https://www.agner.org/optimize/optimizing_assembly.pdf

jdemel replied:
The previous code structure basically tried to prefetch the next chunk of data as well. But in a dangerous way. Now, this should be more secure and the code should be more verbose. Also, this should clarify intention.
We had the "should we use `__VOLK_PREFETCH`" discussion in other threads as well. My conclusion was, to have it as long as it doesn't hurt.

michaelld replied:
> We had the "should we use `__VOLK_PREFETCH`" discussion in other threads as well. My conclusion was, to have it as long as it doesn't hurt.

Yes! Let's make reasonable efforts to not do it dangerously, and only do it if it doesn't hurt performance. The cache manager will auto-prefetch the data sometimes, but not others. Some compilers (& OSs for that matter) will support user-based prefetch, while others don't. We do our best to keep it safe & performance enhancing!

rear1019 reviewed (approved):
LGTM (ignoring my nitpicking about the usage of ``prefetch``).

michaelld commented:
> For reliable results I run `volk_profile -n -R square_dist_scalar_mult -v 1024 -i 10000000`. Or even more
> iterations. What's your hardware setup? On my machines a_avx2 is slightly faster than a_avx. The unaligned versions are way slower.

MacBook Pro 2019 15", Intel Core i9 @ 8-core @ 2.3 GHz
```
% volk_profile -n -R square_dist_scalar_mult -v 1024 -i 10000000

Warning: this IS a dry-run. Config will not be written!
RUN_VOLK_TESTS: volk_32fc_x2_s32f_square_dist_scalar_mult_32f(1024,10000000)
a_avx2 completed in 1433.77 ms
a_avx completed in 2554.14 ms
a_sse3 completed in 2331.53 ms
generic completed in 1806.33 ms
u_avx2 completed in 1492.95 ms
Best aligned arch: a_avx2
Best unaligned arch: u_avx2
Warning: this was a dry-run. Config not generated
```

michaelld commented:
CI still looks good. Not sure why the BB versions fail, but they have nothing to do with this PR.

Is everyone satisfied with the current state of this PR?

jdemel commented:
It seems like everyone is satisfied. Can you merge this PR @michaelld ?
